### PR TITLE
Pr/cleanup keystroke

### PR DESCRIPTION
**Pull request by uli42**
some improvements to keystroke handling.

### CODE REVIEW

**Comment by uli42** [comment]
I have updated this PR to keep the original functionality. modifier-only shortcuts are not supported.

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c [hunk 1]
```diff
@@ -249,6 +249,7 @@ static void parse_keystroke_file(void)
   char *homefile = "/.nx/config/keystrokes.cfg";
   char *etcfile = "/etc/nxagent/keystrokes.cfg";
 
+  /* FIXME: we do not want to have x2go references int the code */
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: typo: s/int/in/

And I partially disagree here. We have the X2Go logo hard-coded into nxagent, too. Having hard-code X2Go related code section was a pre-requisite for merging the former x2goagent fork with the original nxagent code base.

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c [hunk 2]
```diff
@@ -249,6 +249,11 @@ static void parse_keystroke_file(void)
   char *homefile = "/.nx/config/keystrokes.cfg";
   char *etcfile = "/etc/nxagent/keystrokes.cfg";
 
+  if (nxagentKeystrokeFileParsed)
+    return;
+
+  nxagentKeystrokeFileParsed = True;
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: Don't set this to true, if the parsing has not happened so far.

It actually should only be set to true, if the parsing of one of the potential files has _really_ been successful.

Comment (sunweaver) [reply nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: Getting the code design now after reading more of the code context. The nxagentKeystrokeFileParsed is actually more a Boolean for nxagentKeystrokeFileParsingAttempted. It is saying nothing about its success, only that it happened.

We should add a comment about this near the nxagentKeystrokeFileParsed = True assignment, I'd say.

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c [hunk 3]
```diff
@@ -395,6 +395,8 @@ static enum nxagentSpecialKeystroke find_keystroke(XKeyEvent *X)
   int keysyms_per_keycode_return;
   struct nxagentSpecialKeystrokeMap *cur = map;
 
+  /* FIXME: we do late parsing here, this should be done at startup,
+     not at first keypress! */
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: So why not move it over to Init.c?

**Review by sunweaver — CHANGES_REQUESTED** [review]
@uli42: I added some comments / requests. Please check, revisit and give feedback. Thanks.

**Comment by sunweaver** [comment]
Hi Uli,

On  Sa 11 Mär 2017 09:44:35 CET, Ulrich Sibiller wrote:

> some improvements to keystroke handling.
>
> Before the code did not accept keystrokes that did not match a  
> Symbol (e.g. when the config file did not define a key but modifiers  
> only for a keystroke). This code accepts it to allow for  
> modifier-only keystrokes. But this does not work properly because of  
> X11 limitations so this will probably change again.
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ArcticaProject/nx-libs/pull/392
>
> -- Commit Summary --
>
>   * Keystroke.c: use Bool type where appropriate
>   * Keystroke.c: rework read_binding_from_xmlnode()
>   * Keystroke.c: fix code formatting, add FIXME
>   * Keystroke.c: simplify nxagentCheckSpecialKeystroke
>   * Keystroke.c: handle parsing state in parse_keystroke_file()
>   * Keystroke.c: reorder parse_keystroke_file
>   * Keystroke.c: Boolify nxagentCheckSpecialKeystroke
>   * Keystroke.c: Boolify read_binding_from_xmlnode
>   * Keystroke.c: use symbols instead of integers
>   * Keystroke.c: Simplify read_binding_from_xmlnode
>   * Keystroke.c: add another FIXME
>   * Keystroke.c: init struct
>   * Keystroke.c: improve TEST output
>   * Keystroke.c: print keystroke name in TEST mode
>   * Keystroke.c: arrange switch_all_screens and fullscreen close together
>
> -- File Changes --
>
>     M etc/keystrokes.cfg (2)
>     M nx-X11/programs/Xserver/hw/nxagent/Keystroke.c (264)
>     M nx-X11/programs/Xserver/hw/nxagent/Keystroke.h (32)
>
> -- Patch Links --
>
> https://github.com/ArcticaProject/nx-libs/pull/392.patch
> https://github.com/ArcticaProject/nx-libs/pull/392.diff

This PR also introduces a compiler warning, I think. Haven't seen it so far...

```
Keystroke.c: In function ‘parse_keystroke_file’:
Keystroke.c:390:15: warning: passing argument 1 of ‘fprintf’ from  
incompatible pointer type
        fprintf("XML parsing for %s failed\n", filename);
                ^
In file included from ../../include/misc.h:198:0,
                  from ../../include/screenint.h:51,
                  from Keystroke.c:29:
/usr/include/stdio.h:356:12: note: expected ‘struct FILE * restrict’  
but argument is of type ‘char *’
  extern int fprintf (FILE *__restrict __stream,
             ^
```
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by Ionic** [comment]
Yeah, this totally won't work. Either `fprintf(stderr, ...)` or `printf(...)`.

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c (hunk 1 above)
Comment (uli42) [reply nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: OK, I thought we wanted to get rid of that completely. Dropping that comment for now.

**Comment by uli42** [comment]
I have done (or dropped) the requested changes and added some more commits. Please review once more.

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c [hunk 4]
```diff
@@ -204,24 +205,24 @@ static Bool read_binding_from_xmlnode(xmlNode *node, struct nxagentSpecialKeystr
     }
     else if (strcmp((char *)attr->name, "key") == 0)
     {
-      if (strcmp((char *)attr->children->content, "0") != 0 && strcmp((char *)attr->children->content, "false") != 0)
-	newkm.keysym = XStringToKeysym((char *)attr->children->content);
-
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: Why did you remove the above if clause here? Did you do it intentionally? Or is it an editing artifact?

**Review by sunweaver — APPROVED** [review]
One last question (about the removed if clause). Otherwise approved.

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c (hunk 4 above)
Comment (uli42) [reply nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: I am pretty sure it was me who added that code (copy from some lines below), but it is not neccessary. I had added it to allow for "key=0" in my fruitless attempts to implement modifier-only keystrokes.

verified: was added via 7a61c62cd6aaabe40d788d3da7f1b736107fbddb

**Pull request merged**